File: projects/MusicGame/Music-Game-main/src/Background.java
import javax.swing.*;
import java.awt.Color;
import java.awt.Graphics;
import java.util.ArrayList;
import java.util.Random;

public class Background extends JComponent {
    // Constants for the width and height of the game area
    private static final int WIDTH = 800;
    private static final int HEIGHT = 600;
    
    // The player object
    private Player player;
    // List of all paddles in the game
    public static ArrayList<Paddle> paddleList;
    // Label to display the player's score
    private JLabel scores;

    // Timer for game updates
    private Timer timer;
    // Flag to indicate if the game is over
    private boolean gameOver;

    // RGB values to adjust background color as game progresses
    private int red = 245;
    private int green = 245; //Default RGB for beige color
    private int blue = 220;

    // List of power-ups
     private ArrayList<PowerUp> powerUps = new ArrayList<>();

    // Constructor
    public Background(JLabel scores) {
        this.scores = scores;
    
        // Initialize player and paddles
        player = new Player(WIDTH / 2 - Player.SIZE / 2, 0, 0, 1);
        paddleList = new ArrayList<>();
        initializePaddles();

        // Start the game update timer
        timer = new Timer(10, e -> tick());
        timer.start();
    }

    // Returns the list of paddles
    public ArrayList<Paddle> getPaddleList() {
        return paddleList;
    }

    // Initialize paddles at random positions
    public void initializePaddles() {
        Random random = new Random();
        int numPaddles = 15; 
        int spacing = HEIGHT / (numPaddles + 1); // calculate the spacing between the paddles
        Paddle pad = new Paddle(player.getX(), player.getY());
        paddleList.add(pad);
     
        for (int i = 0; i < numPaddles; i++) {
            int y = spacing * (i + 1) - Paddle.HEIGHT / 2; // calculate the y-coordinate of the paddle
            int x = random.nextInt(WIDTH - Paddle.WIDTH); // generate a random x-coordinate
            Paddle paddle = new Paddle(x, y);
            paddleList.add(paddle);
        }
    }

    // Game logic
    void tick() {
        player.move();
        for (Paddle p : paddleList) {
            p.moveDown(HEIGHT, WIDTH); // move the paddle downwards
        
            if (player.isCollidingWithPaddle(p)) {
                player.bounce(); // make the player bounce if it lands on a paddle
                player.increaseScore(); // increase the player's score

                Sound.playNote(player);
            }
        }
        
        repaint();

        scores.setText("Score " + player.getScore());
    }

    @Override
    public void paintComponent(Graphics g) {
        g.setColor(new Color(red,green,blue)); // Light Beige
        g.fillRect(0, 0, WIDTH, HEIGHT);
    
        super.paintComponent(g);
        player.draw(g);

        for (Paddle p : paddleList) {
            p.draw(g);
        }
        for (PowerUp powerUp : powerUps) {
            powerUp.draw(g);
        }
    }

    // Reset game state
    public void reset() {
        player = new Player(WIDTH / 2, 0, 0, 1);
        paddleList.clear();
        initializePaddles(); // Reinitialize paddles
        resetPowerUps(); // Clear power-ups
        resetColor();
        repaint(); // Repaint the JFrame
    }

    // Get the player object
    public Player getPlayer() {
        return player;
    }

    // Pause game
    public void pauseGame() {
        timer.stop();
    }

    // Resume game
    public void resumeGame() {
        timer.start();
    }
    
    // Check if the game is over
    public boolean isGameOver() {
        return gameOver;
    }

    // Set the game over state
    public void setGameOver(boolean gameOver) {
        this.gameOver = gameOver;
    }

    public ArrayList<PowerUp> getPowerUps() {
        return this.powerUps;
    }

    public void setPowerUps(ArrayList<PowerUp> powerUps) {
        this.powerUps = powerUps;
    }
    
    public static int getFrameWidth(){
        return WIDTH;
    }

    public static int getFrameHeight(){
        return HEIGHT;
    }

    // In the Background class
    public Paddle getPaddleCollidingWithPlayer(Player player) {
    for (Paddle paddle : getPaddleList()) {
        if (player.isCollidingWithPaddle(paddle)) {
            return paddle;
        }
    }
    return null;
    }

    /* Old dimColors no longer functional with new gameprogression
    //Set RGB values if valid
    public void dimColors(int dimmer){
        if (red >= dimmer && green >= dimmer && blue >= dimmer){
            red -= dimmer;
            green -= dimmer;
            blue -= dimmer;
        }
        else{
            red = 0;
            green = 0;
            blue = 0;
        }
    } */

    public void setColors(int r, int g, int b){
        red = r;
        green = g;
        blue = b;
    }

    public void resetColor(){
        red = 245;
        green = 245; //Default RGB for beige color
        blue = 220;
    }
    public void resetPowerUps() {
        powerUps.clear();
        repaint();
    }

    
}

File: projects/MusicGame/Music-Game-main/src/Background.js
import { Player } from './Player.js';
import { Paddle } from './Paddle.js';
import { PowerUp } from './PowerUp.js';
import { Sound } from './Sound.js';

class Background extends JComponent {
    // Constants for the width and height of the game area
    static WIDTH = 800;
    static HEIGHT = 600;

    constructor(scores) {
        super();
        this.scores = scores;

        // Initialize player and paddles
        this.player = new Player(Background.WIDTH / 2 - Player.SIZE / 2, 0, 0, 1);
        Background.paddleList = [];
        this.initializePaddles();

        // Start the game update timer
        this.timer = setInterval(() => this.tick(), 10);
        this.gameOver = false;

        // RGB values to adjust background color as game progresses
        this.red = 245;
        this.green = 245; // Default RGB for beige color
        this.blue = 220;

        // List of power-ups
        this.powerUps = [];
    }

    // Returns the list of paddles
    getPaddleList() {
        return Background.paddleList;
    }

    // Initialize paddles at random positions
    initializePaddles() {
        const random = new Random();
        const numPaddles = 15;
        const spacing = Background.HEIGHT / (numPaddles + 1); // calculate the spacing between the paddles
        const pad = new Paddle(this.player.getX(), this.player.getY());
        Background.paddleList.push(pad);

        for (let i = 0; i < numPaddles; i++) {
            const y = spacing * (i + 1) - Paddle.HEIGHT / 2; // calculate the y-coordinate of the paddle
            const x = Math.floor(Math.random() * (Background.WIDTH - Paddle.WIDTH)); // generate a random x-coordinate
            const paddle = new Paddle(x, y);
            Background.paddleList.push(paddle);
        }
    }

    // Game logic
    tick() {
        this.player.move();
        for (const p of Background.paddleList) {
            p.moveDown(Background.HEIGHT, Background.WIDTH); // move the paddle downwards

            if (this.player.isCollidingWithPaddle(p)) {
                this.player.bounce(); // make the player bounce if it lands on a paddle
                this.player.increaseScore(); // increase the player's score

                Sound.playNote(this.player);
            }
        }

        this.repaint();

        this.scores.setText("Score " + this.player.getScore());
    }

    paintComponent(g) {
        g.setColor(new Color(this.red, this.green, this.blue)); // Light Beige
        g.fillRect(0, 0, Background.WIDTH, Background.HEIGHT);

        super.paintComponent(g);
        this.player.draw(g);

        for (const p of Background.paddleList) {
            p.draw(g);
        }
        for (const powerUp of this.powerUps) {
            powerUp.draw(g);
        }
    }

    // Reset game state
    reset() {
        this.player = new Player(Background.WIDTH / 2, 0, 0, 1);
        Background.paddleList = [];
        this.initializePaddles(); // Reinitialize paddles
        this.resetPowerUps(); // Clear power-ups
        this.resetColor();
        this.repaint(); // Repaint the JFrame
    }

    // Get the player object
    getPlayer() {
        return this.player;
    }

    // Pause game
    pauseGame() {
        clearInterval(this.timer);
    }

    // Resume game
    resumeGame() {
        this.timer = setInterval(() => this.tick(), 10);
    }

    // Check if the game is over
    isGameOver() {
        return this.gameOver;
    }

    // Set the game over state
    setGameOver(gameOver) {
        this.gameOver = gameOver;
    }

    getPowerUps() {
        return this.powerUps;
    }

    setPowerUps(powerUps) {
        this.powerUps = powerUps;
    }

    static getFrameWidth() {
        return Background.WIDTH;
    }

    static getFrameHeight() {
        return Background.HEIGHT;
    }

    getPaddleCollidingWithPlayer(player) {
        for (const paddle of this.getPaddleList()) {
            if (player.isCollidingWithPaddle(paddle)) {
                return paddle;
            }
        }
        return null;
    }

    setColors(r, g, b) {
        this.red = r;
        this.green = g;
        this.blue = b;
    }

    resetColor() {
        this.red = 245;
        this.green = 245; // Default RGB for beige color
        this.blue = 220;
    }

    resetPowerUps() {
        this.powerUps = [];
        this.repaint();
    }
}

export { Background };